### ISSUE TITLE
Post 0.8 ideas

### ISSUE THREAD

**Issue by fafhrd91** [actix/actix#224]
some ideas

- Split tokio
- Re-evaluate `ActorFuture` types
- Review `Response` type, i think it could be replaced with `IntoFuture`

@actix/contributors

**Comment by najamelan** [comment]
@actix/contributors 

I would propose trying to get actix to a minimal lib. Just keep the traits: Actor, Message, Context with mailbox, ... But remove things like resolver, actorfuture, ... Put them in separate crates which use actix. That way we can try to bring actix to v1 by cleaning everything up, making sure there is clear documentation with examples for every public api, unit test everything, adhere to rust api guidelines, make sure the book is complete, porting to async-await, ...

It would allow also to be less strict on response type, and we can think about actorfuture later when it is in a different crate.

Minimal, performant, actor model.    

> Split tokio

Currently tokio is only required for timer (which is innocent I think) and for resolver which we can move to a different crate. Then there is 
actix-rt which needs some cleaning up. I made[ a proof of concept](https://github.com/najamelan/actix/tree/wasm-runtime/examples/wasm) which runs on WASM, and does not use tokio as a runtime. For basic usage, all that was needed was the possibility to spawn futures. 

In itself an actor model does not need to be tied to IO.

**Comment by fafhrd91** [comment]
I am not sure about removing actorfuture. ActorFuture is a point of actix, it gives you access to the state.

I have an idea how to have similar behavior without actorfuture. But it requires some research and time which I don’t have at the moment. I can explain If anybody wants to try it.

**Comment by najamelan** [comment]
So I'm having a look at ActorFuture. From the docs:
```rust
impl Message for SomeMessage { type Result = (); }

impl Handler<DeferredWork> for OriginalActor
{
	type Result = ResponseActFuture< Self, (), Error >;

	fn handle( &mut self, _msg: Refresh, ctx: &mut Context<Self> ) -> Self::Result
	{
		let send_to_other = 

			self.other_actor_addr
			.send   ( OtherMessage::new() )
			.map_err( Error::from         )
		;

		let update_self =

			wrap_future::< _, Self >( send_to_other ).map( |result, actor, _ctx|
			{
				// Actor's state updated here
				//
				actor.inner_state.update_from( result );
			})
		;

		// return the wrapping future
		//
		Box::new(update_self);
	}
}
```
Do I get it right that the existence of actorfuture is because `self` can't be accessed in the closure? If so, why not?
```rust
MyActor
{
	inner_state: Rc<RefCell< State >> ,
}
```
then use `clone()` and `borrow()` or `borrow_mut()`, which will work just fine?

**Comment by fafhrd91** [comment]
in that case you need actors at all? actor is just object with mpsc channel

**Comment by najamelan** [comment]
For me actors are a software design model. They allow loosely coupled design. When messages can be serialized, they can live anywhere. Communication is not tied to having references, or pointers, just the abstract notion of address which might be implemented in different ways. They allow me to think about each component almost entirely in isolation. Just messages come in and go out. Actors are capabilities as well. You cannot send anything unless you have the address. 

The concept is simple and elegant. Mainly pushing the concept of "loose coupling, tight cohesion" as far as possible.  

I don't really see why the use of `Rc` and `RefCell` would change anything here... but the question is, if you can use `Rc` and `RefCell`, why use `ActorFuture`?

**Comment by DoumanAsh** [comment]
@fafhrd91 Important idea, we need to allow dropping out of  `Mailbox::poll` when we reach some sort of limit, to allow processing other futures.

I'll need to think about it again and find out why it didn't work last time I tried

**Comment by semtexzv** [comment]
ActorFuture is currently neccessary in core, but in the future it could be possible to replace it with Async-await style. I've been doing some experiments here: https://github.com/semtexzv/ak

**Comment by kpp** [comment]
Sure but there is ` u32::from_le_bytes(make_array(&self.buf[0..])),` in this repo. Use one style

**Comment by najamelan** [comment]
@semtexzv I am also running some experiments. This seems to [run just fine](https://github.com/thespis-rs), only it's slow, so I'm researching the performance. It might be the std channels involved. Making a crossbeam async await channel today if that's where it drags it's feet. The iface crate has only traits defining the interface, the impl crate is some experimental implementation. The idea is that an address guarantees that it knows how to deliver the message, so someone could implement an addresss type that boils down to an IP address and PORT, that could be serialized and send accross the wire, or use libp2p or DHT... Separating the interface from the impl avoids locking the users in an impl...

**Comment by fafhrd91** [comment]
@najamelan check [actix-remote](https://github.com/actix/actix-remote) I played a bit with distributed actors

**Comment by najamelan** [comment]
@fafhrd91 Yes, I had a look at it. I see it is rather complex, with nodes registering and so. I am currently more thinking about the interface. Then many implementations can be written doing different things. Currently for me it seems good that one can obtain different address types for the same actor. Thus in my experiment Address is a trait. Thus you can get a local in process address for high performance, but also, you could get a WebSocket address, which would open a ws server on a port, and the address can serialize as ip:port, now you can deserialize it on a different machine, call send on it and it will know how to send the message to the remote actor. Keeping it simple ;) Of course, this would require a port per actor that can be joined remotely, so other implementation might make it so that only one port is needed per process and use some dispatching mechanism... Some people will figure out how to write mailbox, address, ... for no_std, for wasm, on a single threaded machine, because an interface allows this flexibility. And if there is 1 interface, actors from these different implementation can still talk to one another.

**Comment by najamelan** [comment]
let's have some fun and make multiactor for [multiformat](https://github.com/multiformats/multiformats)

**Comment by najamelan** [comment]
Ok, so over the last month I have been working on how I imagine an Actor Model in Rust. I think it's coming along quite nicely and want to [share it here](https://github.com/thespis-rs/thespis). The readme explains a bit how it's different from Actix and what the design goals are. The performance problems mentioned above are long gone and it's actually quite fast. (on my laptop, it takes 160ns to send a message to an actor in the same thread, opposed to 10ns for a normal async method call and 370ns with actix `do_send`). I have yet to do performance optimization. The difference might be because right now I use unbounded channels, which might be faster than bounded.

I would like to gradually bring it to publishable quality over the next few months.

@fafhrd91 I hope you don't mind if I publish this on crates, it does take inspiration from Actix and has a similar API (traits Actor, Message, Recipient, Handler, Envelope, Mailbox ...)

One thing we can all think about is that in thespis I try to make an interface (only traits) that express the actor model. Now different libraries can make different implementations, yet could be compatible. If over time we could all agree on such interface, it would be cool! and all actors even from different libs can talk to eachother.

**Comment by semtexzv** [comment]
I would like to revisit this issue, as I am currently in the process of migrating actix to std::future. 

The issue with ActorFuture is the general pollution of codebase and complexity. But its main goal is to provide an access to the actor state. This worked fine when we used the same combinators as used by normal `Future` implementations and got the access to the actor state 'for free'.

Now, in the land of async-await, this does not work that great. Creating ActorFuture and chaining them using combinators works, but you can't simply await on them. In general, the code looks like this:

```rust
impl Handler<Msg> for Actor {
    type Result = ();

    fn handle(&mut self, msg: IngestUpdate, ctx: &mut Context<Self>) {
        sync_action();
        sync_action();
        let future = async_action()
            .into_actor(self)
            .then(|res, this, ctx| {
                  sync_action();
                  sync_action();
                async {
                    let a = async_action().await;
                    let b = async_action().await;
                    (a,b)
                }.into_actor(this)
            })
            .then(|res,this,ctx| {
                  sync_action(res);
                  sync_action(res);
            });
        
        // either spawn or return the future as actorFuture
    }
}

```

Which I don't really like from ergonomics standpoint and issues with lifetimes.
I am trying to get to something line this
```rust
impl Handler<Msg> for Actor {
    type Result = ();

    fn handle(&mut self, msg: IngestUpdate, ctx: &mut Context<Self>) {
        sync_action();
        sync_action();
        async {
            let (res,this,ctx) = ctx.suspend(this,async_action()).await;
            sync_action(ctx);
            sync_action(ctx);
            let (res,this,ctx) = ctx.suspend(this,async_action()).await;
            let (this,ctx) = if a {
                let (res,this,ctx) = ctx.suspend(this,async_action1()).await;
                (this,ctx)
            } else {
                let (res,this,ctx) = ctx.suspend(this,async_action2()).await;
                (this,ctx)
            };
        }
        // either spawn or return the future as actorFuture
    }
}
```

Basically, manual transferring of Self + Context references, that forces references to state to be invalidated, and creation of new Self + Context references upon resuming execution from a suspended future.

This would require some kind of `Cell` wrapper, that would internally contain a pointer, would be transferred by ownership into method similar to this(this would be method on context):
```rust
pub fn suspend(self : OwnedCell<Self>,  act : OwnedCell<A>, fut : F) 
-> Future<(OwnedCell<Self>, OwnedCell<A>, R) where F: Future<Output=R>>
```
Or: even better with a Wrapper type that implements context methods and derefs to an actor:
```rust
pub fn suspend(self : ContextRef<A> fut : F) 
-> Future<(ContextRef<Self>, R) where F: Future<Output=R>>
```

**Comment by fafhrd91** [comment]
Looks interesting. Would it be possible to ha for message concurrently?

**Comment by semtexzv** [comment]
I believe so, the suspend method would transfer ownership of  the context and actor back into the Context, which means that all references are invalid after that point.

In the current system, the only way future ends up on in the actors context is by spawn and wait methods. Even if a future is returned from the handler, the impl of `MessageResponse` spawns it and then sends a result over a channel. Current impl for spawn method only accepts `ActorFuture`, 

Now, in new design, the suspend method would internally acccess the Context through the `ContextRef`, would add spawn a future which would execute the contained inner future, and resolve into provided ContextRef.

There is an issue with unsafety, but if we can guarantee that spawned futures do not outlive the actor context, and don't escape it, I think, we can get by with a RefCell, or even a simple pointer to the context.


Take a look at : 
https://github.com/semtexzv/ak/blob/master/src/context.rs
 and using it like:
https://github.com/semtexzv/ak/blob/fbe2b4c22655b33571248d9a7a36f2a4aba3409d/src/lib.rs#L54

These were my experiments, in which I transferred the ownership of `Box<Context>` through the handler, which meant that every `.suspend` would re-ener context, and transfer its ownereship to itself. 

This did not work so well, because it meant I had to return the `Box<Context>` upon exiting the handler every time, and it screwed up the ergonomics of the whole thing.

But using a wrapped unsafe pointer, and keeping the 'ownership' of the context to the ContextFuture, which is polled by executor like actix does, with combination of my approach would work wonders from ergonomic standpoint.

**Comment by semtexzv** [comment]
After experimenting with async-await on Actix-0.8, I've come to several conclusions:

1. Async actors should Support both Synchronous and asynchronous handlers, but the asynchronous should be usable with async-await.
2. This introduces 2 "environments". The synchronous one, where we are allowed to have references to `Self`, and `Self::Context`, and asynchronous one, where we can await on futures. 
3. The main problem is delineating these 2 environments. The design proposed above has problems, because it mixes these 2 designs.
4. Sucessful design will have to provide a way to separate these environments.

So, I propose following design.

Add a `LocalAddr<Self>` type. This internally just wraps an `Addr<Self>` at first. The code can only obtain it from Context<Self>, and can't sent it to other actors.

The `AsyncAddr` will have `pub async fn invoke_sync(&self, f : F) -> R` method, which will take an `FnOnce(&mut Self, &mut Self::Context) -> R` and return the R value.
 

```rust
impl LocalAddr<A : Actor> {
    async fn invoke<F,R>(&self, f: F) -> R
        where F: FnOnce(&mut A, &mut <A as Actor>::Context) -> R + Send + 'static,
          R: Send + 'static;
}

```
This function could be used to enter the synchronous environment safely from the asynchronous one. 

To go from synchronous environment into asynchronous one, we just have to start an async block.
However, It would be beneficial, to implement a method on AsyncContext<A>, which would consume references to Self  and Ctx, and would accept async closure

```rust
impl Context<A> {
    fn async<F, Fut, R>(&mut self, &mut a, f : F ) -> Response<A, R>
        where F: FnOnce(LocalAddr<A>) -> Fut,
                   Fut: Future<Output=R> {


        Response::fut(wrap_future(f(self.local_addr()))
    }

}
```

So that resulting Handler impl would look like this :

```rust
impl Handler<Msg> for MyActor {
    type Result = Response<Self, ()>;
    fn handle(&mut self, msg: Msg, ctx: &mut Self::Context) -> Self::Result {
        println!("Sync environment");
        let fut = self.db.load_all();

        ctx.async(self, move |addr| move {
            let x = fut.await;

            let modified = addr.invoke(|this,ctx|{
                this.value += 1;
                this.db.load_one(this.value)
            });

            let y = modified.await;
        })
    }
}
```

Problems :
1.  Verbose - but could be better than the alternatives.
2. Performance - Currently this sends a new message to Self, and waits for response, could be optimized to insert a message into the `ContextFut` wait queue, or similar one in order to reduce latency and overhead, and drop the `'static` lifetiem bounds
3. Lifetimes- returing futures generated by synchronous code is currently a problem, because we require that the messages must be 'static. If similar approach is implemented natively in `ContextFut`, without requiring sending messages, maybe we can drop some of these requirements.


Let e  know what do you think, and how could we improve this. I am constantly looking for solution to this problem.

**Comment by uloskae** [comment]
any updates :(?

**Comment by robjtede** [comment]
> any updates :(?

@uloskae What point are you most interested in getting an update on?

Regardless, I think this should be closed since we're now on v0.10 and ideas that deserve attention can be made into specific issues.
